configurable resolution limit

wbruna opened the pull request:
This replaces the current hardcoded resolution limit (#1567) with a config variable.

* I kept the current 'clamping' limit as is, since it has a different effect, and people may be used to that functionality;
* I placed the config field near the clamping one, since they both limit image size / memory usage, but instead of messing around directly with the row indexes, I replaced them with a 'current row' variable; it should make it easier to add or reorder fields. I kept that change separate to ease reviewing;
* the only common unit for an image area is 'megapixels', so I used that, although we need fairly low values in this case. I chose the 0.7MP limit to place it just above the current maximum image size for SD 1.5 (832x832), but it will restrict SDXL generations a little bit more by default (on my own card I can reach 0.72 without allocation errors).

@LostRuins , please feel free to do or request any changes!

wbruna committed:
make image size limit configurable

wbruna committed:
refactor image gen configuration screen

LostRuins commented:
Hm I'm trying to understand the purpose of this flag. 
So right now with SD, there are 2 main limits that are hit with big images:
- The VRAM/Memory use increases quadratically, leading to OOM situations
- Issues with the implementation leading to crashes beyond certain dimensions

The current `--sdclamped` solves this issue by ensuring that you never cross both scenarios. Because trying to generate a `16x8192` or `8x12000` image will absolutely go wrong, and so will generating a `3000x3000` images.

I guess if I understand it correctly, you're trying to create a scenario where if `1024x1024` is permitted, then `512x2048` and `256x4096` is also permitted? But this cannot be reliably guaranteed.  Or am I misunderstanding the feature

wbruna commented:
Just to make it clearer: the main purpose is being able to generate higher resolution non-square images without risking server crashes.

> The current --sdclamped solves this issue by ensuring that you never cross both scenarios.

Yes, as a safety feature, I agree the current approach is _enough_: the limit can be set low enough to ensure the server does not crash. It's just too low of a limit for landscape or portrait images, since OOM conditions depend mainly on the image area. My card can eg. generate 1024x576 widescreen images even without VAE tiling, but currently I couldn't allow it safely because the only available limit is a biggest dimension cap: I'd need to set it to 832, because a 896x896 request goes bad already.

>  Because trying to generate a 16x8192 or 8x12000 image will absolutely go wrong, and so will generating a 3000x3000 images.

Just a nitpick, but: a 0.5MP limit, together with the unavoidable limitation of all sides being multiples of 64, would cap the resolution to something like 64x8192 - and command-line sd.cpp renders that just fine. Sure, we could argue if the result is useful or good, but that's not a safety issue 🙂

wbruna committed:
fix resolution limits and keep dimensions closer to the original ratio

wbruna committed:
use 0.0 for the configured default image size limit

This prevents the current default value from being saved into the
config files, in case we later decide to adopt a different value.

LostRuins commented:
I think that accepting resolutions as megapixels is not the best, but I don't know of a better way to represent it. Megapixels become ugly floating point numbers and it's hard for a user to estimate how many their device can actually handle. 

Ideally it should instead be a boolean toggle from the `--sdclamped`, where is `sd clamp strict mode (default)` is set, then sdclamp will apply on ALL dimensions (e.g. clamp 1024 ensures width < 1024 and height < 1024).

If strict mode is disabled, then the clamp internally uses megapixels instead. The default clamp will become `1024x1024` = 1.05 megapixels. And THEN you apply your logic based on that. 

**Thus:**
`--sdclamped 1024` + `--sdclamppermissive` (example name)  would allow both `1024x1024` and `512x2048` as well.

 It makes it a lot simpler from the user POV. 

I guess this may be useful as an advanced approach, I would probably make it off by default (relying on the old clamping behavior). 

If that name is too silly, we can do and integer toggle `--sdclampedstrict 1 or 0` instead. Though a boolean allowtrue flag feels neater.

wbruna commented:
> I think that accepting resolutions as megapixels is not the best, but I don't know of a better way to represent it. Megapixels become ugly floating point numbers and it's hard for a user to estimate how many their device can actually handle.

Just thinking out loud a bit: regardless of the size limit option, what about displaying a "memory cost" value on the sdui screen, updated when the user changes the image dimension sliders? Something similar to the Horde's kudos calculation, but browser-side only. That could be helpful for low-end devices: for instance, if above a certain value the GPU needs to swap out memory, the generation becomes much slower, and that cost number would give a direct feedback about that possibility.

> Ideally it should instead be a boolean toggle from the --sdclamped, where is sd clamp strict mode (default) is set, then sdclamp will apply on ALL dimensions (e.g. clamp 1024 ensures width < 1024 and height < 1024).

Yeah, that could work. I have seen the term "lenient" on a similar context (no idea if that sounds better, though; I'm far from a native speaker).

In this case, what would you think about decoupling or tweaking the step count limit? I don't mind it that much right now (40 is already above my patience level 🙂), but if we're counting on an existing config flag for this limit, we should also take a look at what else it does.

If the step number limitation intends to clamp the generation time/processing cost, it's not that good of an estimate for many cases: eg. for img2img, the true step count is something like step_count * strength, so a 0.2 strength img2img would be limited to ~8 steps. Also, a CFG 1 generation needs half the processing. Limiting by an abstract "processing cost" number (step_count * strength * (2 if cfg>1 else 1) ) could be much more accurate.

I tested a few generations with an SDXL model, at step count 10, not including the VAE time:

|resolution | guidance| mode         |actual step count|step speed|gen time|
|-----------|--------:|-------------:|-----------------|---------:|-------:|
| 1024x1024 |     1.5 | txt2img      |              10 | 4.03s/it | 40.42s |
| 1024x1024 |     1.5 | img2img 0.5  |               6 | 4.03s/it | 24.64s |
| 1024x1024 |     1.0 | txt2img      |              10 | 2.05s/it | 21.03s |
| 1024x1024 |     1.0 | img2img 0.5  |               6 | 2.05s/it | 12.43s |
|  512x1024 |     1.5 | txt2img      |              10 | 1.67s/it | 16.93s |
|  512x1024 |     1.5 | img2img 0.5  |               6 | 1.67s/it | 10.21s |
|  512x1024 |     1.0 | txt2img      |              10 | 1.19it/s |  8.81s |
|  512x1024 |     1.0 | img2img 0.5  |               6 | 1.19it/s |  5.18s |
|  512x512  |     1.5 | txt2img      |              10 | 1.06it/s |  9.32s |
|  512x512  |     1.5 | img2img 0.5  |               6 | 1.06it/s |  5.73s |
|  512x512  |     1.0 | txt2img      |              10 | 2.10it/s |  4.81s |
|  512x512  |     1.0 | img2img 0.5  |               6 | 2.24it/s |  2.96s |

> I guess this may be useful as an advanced approach, I would probably make it off by default (relying on the old clamping behavior).

You mean the per-model-type limit by default? Sure, I'll give it a try.

LostRuins commented:
Step count limit is for shared server use to prevent abuse. Nothing to do with crashing. It can be overridden by the user with an adapter config (though its a slightly tedious process).

wbruna commented:
> Step count limit is for shared server use to prevent abuse. Nothing to do with crashing.

I asked because the --sdclamped option is also applying a step count limit:

https://github.com/LostRuins/koboldcpp/blob/v1.93.1/koboldcpp.py#L1598
```python
    if args.sdclamped:
        sample_steps = (40 if sample_steps > 40 else sample_steps)
```
I believe you are referring to a separate forced_steplimit, coming from an adapter_obj, being applied a bit above that?

wbruna committed:
export image model version when loading

wbruna committed:
restore model-specific default image size limit

LostRuins commented:
Correct, one is a hard limit and the other is customizable (as some users wanted it lower than that)

LostRuins commented:
Ok let's not worry about the step counts for now, let's just keep it simple: add the `--sdclampedlenient` flag (boolean flag).

When off:
`--sdclamped nnn` = allow `min(x=nnn,y=nnn)` resolution limit (current behavior).

When on:
`--sdclamped nnn` = allow `mp=x*y < nnn*nnn` resolution limit (permissive lenient behavior)
Can also add a few failsafes e.g. `x<4y && y<4x` or something, but this is the basic

What do you think? we can expand the steps stuff later.

wbruna commented:
All right.

A little UI bikeshedding: perhaps `--sdclampmode=[strict|lenient]` instead of `--sdclampmode=[true|false]`? Easier to change and keep compatibility with, in case you decide to add/remove a policy later. (I can implement both, so we can see how they look; it's easier for me to do it step-by-step anyway).

LostRuins commented:
sure, we can do that too. so long the the default is same as old behavior (strict)

wbruna committed:
change the image area restriction to be specified by a square side

wbruna commented:
@LostRuins , please take a look at this last approach: I replaced the floating-point megapixels parameter by a plain square image side. This has a direct match to something the user would expect (the biggest square image they can generate), and it can be tested by simply generating a few square images.

I also mention some examples of trade-offs between width and height in the parameter text. The combined restriction would be similar to what some image generation sites do: Dezgo is pretty close to `--sdclamped 680 --sdrestrictsquare 520` , while `--sdclamped 768 --sdrestrictsquare 640` would allow images matching Perchance and Civitai portraits and landscapes while restricting more costly 768x768 square images.

(I'll go ahead and implement the --sdclampmode approach on top of it anyway)

LostRuins commented:
It's somewhat confusing imho. I kind of understand it after reading the description, but it feel very unintuitive to me. 

I leave the choice up to you, but you should only pick one. Either we have the clamped + clampmode, or we have clamped + squaresize, but not both, that would be really confusing I think. 

If we are worried about very extreme aspect ratios I think it just makes sense to clamp them to a comfortable limit. 1:4 to 4:1 is plenty.

wbruna commented:
> I leave the choice up to you, but you should only pick one. Either we have the clamped + clampmode, or we have clamped + squaresize, but not both, that would be really confusing I think.

Oh, I agree. I was only implementing both to make it easier to compare them.

And after a few struggles with the interface, I was able to implement the sdclampmode. I can push it to another branch if you want to take a look, but to be honest, just the flag description makes it more confusing in my opinion, because I have to explain about changing the clamping policy _and_ how the square limit works. So keeping them separate avoids questions about what sdclamping itself does, and gives a bit more functionality.

I'll rebase it on top of concedo_experimental.

wbruna commented:
From a seems-to-be-working POV I think it's ready.

I called the option `sdrestrictsquare`, but I have no strong feelings about that name; could be whatever you think is less confusing. I just feel that using the word 'square' could help understanding what it does (both in the 'square image' and 'the true limit is the square of this number' senses).

LostRuins commented:
You should not be applying the scaling logic in python, all of that should be done on the C++ side. I think it is not a good idea to pass back model name information for the python side to handle - downscaling calculations should be handled where they currently are because the only backend has fully visibility of the model info (in future there are flags like Chroma support, info about vae types, lora types etc). If you need i can refactor that part for you.

wbruna commented:
> You should not be applying the scaling logic in python, all of that should be done on the C++ side. I think it is not a good idea to pass back model name information for the python side to handle - downscaling calculations should be handled where they currently are because the only backend has fully visibility of the model info (in future there are flags like Chroma support, info about vae types, lora types etc). 

Oh, OK. My thought was, it was better to apply it at the Python level, because it''d be easier to give feedback about it in the interface (say, a visual indication that the chosen image size would need rescaling), and easier to test changes. 

About the model metadata, I tried to make it extensible by using a struct. Some kind of structured string (rfc822, or even JSON) would probably be more appropriate, but since I'm currently returning a string for the model version, the code could be adapted rather easily for that.

I can push the config down to the C++ level instead, no problem. But, on that line, I have one observation: I need to push down the sdclamped config too, or at least remove any rounding to 64px from the Python level. The downscale logic needs the original image proportions to make the resulting image as close as possible to the original requested values, and to avoid double-rounding them down to 64px; it can't get pre-rounded dimensions, or it will mess up some ratios.

> If you need i can refactor that part for you.

No need; I actually already have some code for that, from a previous attempt.

LostRuins commented:
alright sure, you can send the original resolutions to the backend and handle from there.

wbruna committed:
move image resolution limits down to the C++ level

wbruna committed:
Revert "export image model version when loading"

This reverts commit fa65b23de31fe0f9706280a038f48c8dd3e7794f.

LostRuins commented:
Do let me know when you're ready for me to review again, thanks!

wbruna commented:
I was chasing down a bug, but it turns out it was a stale leftover from a previous build.

Should be good to go. I left the history as is for now, but I can clean it up once it's ready for merging.

LostRuins commented:
No need to clean up history, eventually when merged it will squash down to a single commit

LostRuins committed:
Linting Fixes:
PY:
- Inconsistent var name sd_restrict_square -> sd_restrict_square_var
- GUI swap back to using absolute row numbers for now.
- fstring fix
- size_limit -> side_limit inconsistency
C++:
- roundup_64 standalone function
- refactor sd_fix_resolution variable names for clarity
- move "anti crashing" hard total megapixel limit always to be applied after soft total megapixel limit instead of conditionally only when sd_restrict_square is unset

LostRuins commented:
Alright did a few minor edits

Linting Fixes:
PY:
- Inconsistent var name sd_restrict_square -> sd_restrict_square_var
- GUI swap back to using absolute row numbers for now.
- fstring fix
- size_limit -> side_limit naming inconsistency
C++:
- roundup_64 standalone function
- refactor sd_fix_resolution variable names for clarity
- move "anti crashing" hard total megapixel limit always to be applied after soft total megapixel limit instead of conditionally only when sd_restrict_square is unset

And i think we are good to merge! Thank you for your contribution

wbruna commented:
> Alright did a few minor edits

All look good to me. Thanks!

> move "anti crashing" hard total megapixel limit always to be applied after soft total megapixel limit instead of conditionally only when sd_restrict_square is unset

Hm... My intention here was to be able to move up the anti-crash limit, because it's likely backend-specific. So a user could move it up if it doesn't affect their particular configuration, and it'd be useful to allow non-developers (or myself, after this change reaches the ROCm builds) to mess around with it to help testing specific backend+model combinations (the help texts are supposed to give the appropriate this-will-break-your-server scares). If it's kept as a true hard limit, the help should probably mention that.

LostRuins commented:
Ahh I see... I thought that the original reason was because you wanted to generate **2048x512** or **1280x720** etc images instead of **1024x1024** images which a simple `--sdclamped 1024` prevented, all these should be possible now. 

I will include additional info in help, but I think that it might not be ideal to link simply setting `sd_restrict_square` to also trigger the safeguards removal, especially since the hardlimit was there previously too. 

Do you often generate images above 1.04 megapixels? Perhaps I can permit that in `--debugmode`, if sdrestrictsquare is explicitly set.

LostRuins committed:
allow unsafe resolutions if debugmode is on

LostRuins commented:
Alright, now with `--debugmode` enabled `--sdrestrictsquare` permits images up to **4 megapixels** which should be more than enough realistically.

Merging now but feel free to add more comments

wbruna commented:
> Ahh I see... I thought that the original reason was because you wanted to generate **2048x512** or **1280x720** etc images instead of **1024x1024** images which a simple `--sdclamped 1024` prevented, all these should be possible now.

That _is_ the main motivation, I'm just incapable of implementing anything without generalizing it a little bit 😀

My main usage is on Vulkan, where I probably can't avoid that limit anyway; but I noticed ROCm doesn't have the same restrictions on the VAE processing, so perhaps it can avoid some of those other limits too.
 
> I will include additional info in help, but I think that it might not be ideal to link simply setting `sd_restrict_square` to also trigger the safeguards removal, especially since the hardlimit was there previously too.
> 
> Do you often generate images above 1.04 megapixels? Perhaps I can permit that in `--debugmode`, if sdrestrictsquare is explicitly set.

Yeah, putting it behind debug mode, or an undocumented env var, would be more than enough for experimenting.

wbruna commented:
On 1.94, by lifting the limit up to 1024 + debug flag, I was able to run a SD1.5 model at 1024x960 with the hipblas backend:

![landscape](https://github.com/user-attachments/assets/171761ff-aa80-46a4-80da-f03f02f2615e)

A full 1024x1024 image crashes with:

```
unet compute buffer size: 8360.93 MB(VRAM)ggml_cuda_compute_forward: SCALE failed
ROCm error: invalid configuration argument
  current device: 0, in function ggml_cuda_compute_forward at ggml/src/ggml-cuda/ggml-cuda.cu:2402
  err
ggml/src/ggml-cuda/ggml-cuda.cu:79: ROCm error
```

So it all points to a hardware/driver/backend memory allocation limit (in my case: around 8G for ROCm, 4G for Vulkan, on a 16G GPU), together with how much memory the sd.cpp implementation needs to allocate in a single buffer for that particular model+resolution.

LostRuins commented:
Yup. The official safe limits are there for a reason, they've been tested to mostly work.

wbruna commented:
And with a modified stable-ui, I was able to go much higher with SDXL, even on Vulkan. This is at 1856x1024; and compared to non-tiled VAE, it wasn't even especially memory-hungry: the card stayed below ~6.3G VRAM (q8_0 model). Perhaps the interface resolution range could be raised a bit?

![landscape_1856x1024](https://github.com/user-attachments/assets/88f4d012-c979-4476-821c-333d913570ac)
